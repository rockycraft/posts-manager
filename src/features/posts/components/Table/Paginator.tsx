import { ButtonGroup, IconButton, Text } from "@chakra-ui/react";
import { HiChevronLeft, HiChevronRight } from "react-icons/hi";

type PaginatorProps = {
	actualPage: number;
	totalPages: number;
	canPreviousPage: boolean;
	canNextPage: boolean;
	onClickPrevious: () => void;
	onClickNext: () => void;
};

export function Paginator({
	actualPage,
	totalPages,
	canPreviousPage,
	canNextPage,
	onClickPrevious,
	onClickNext,
}: PaginatorProps) {
	return (
		<ButtonGroup>
			<Text data-testid="page-detail">{`PÃ¡gina ${actualPage} de ${totalPages}`}</Text>
			<IconButton
				data-testid="previous-button"
				variant="outline"
				onClick={onClickPrevious}
				disabled={!canPreviousPage}
			>
				<HiChevronLeft />
			</IconButton>
			<IconButton
				data-testid="next-button"
				variant="outline"
				onClick={onClickNext}
				disabled={!canNextPage}
			>
				<HiChevronRight />
			</IconButton>
		</ButtonGroup>
	);
}
