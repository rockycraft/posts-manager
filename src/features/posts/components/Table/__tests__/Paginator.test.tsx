import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import { Paginator } from "../Paginator";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { Provider } from "@/components/ui/provider";

Object.defineProperty(window, "matchMedia", {
	writable: true,
	value: vi.fn().mockImplementation((query) => ({
		matches: false,
		media: query,
		onchange: null,
		addListener: vi.fn(), // deprecated
		removeListener: vi.fn(), // deprecated
		addEventListener: vi.fn(),
		removeEventListener: vi.fn(),
		dispatchEvent: vi.fn(),
	})),
});

describe("Paginator component", () => {
	const mockOnClickPrevious = vi.fn();
	const mockOnClickNext = vi.fn();

	const defaultProps = {
		actualPage: 1,
		totalPages: 5,
		canPreviousPage: true,
		canNextPage: true,
		onClickPrevious: mockOnClickPrevious,
		onClickNext: mockOnClickNext,
	};

	beforeEach(() => {
		vi.clearAllMocks();
	});

	it("should render the paginator with correct text", () => {
		render(
			<Provider>
				<Paginator {...defaultProps} />
			</Provider>,
		);

		expect(screen.getByTestId("page-detail")).toBeInTheDocument();
	});

	it("should call onClickPrevious when previous button is clicked", () => {
		render(
			<Provider>
				<Paginator {...defaultProps} />
			</Provider>,
		);

		fireEvent.click(screen.getByTestId("previous-button"));

		expect(mockOnClickPrevious).toHaveBeenCalledTimes(1);
	});

	it("should call onClickNext when next button is clicked", () => {
		render(
			<Provider>
				<Paginator {...defaultProps} />
			</Provider>,
		);

		fireEvent.click(screen.getByTestId("next-button"));

		expect(mockOnClickNext).toHaveBeenCalledTimes(1);
	});

	it("should disable previous button when canPreviousPage is false", () => {
		render(
			<Provider>
				<Paginator {...defaultProps} canPreviousPage={false} />
			</Provider>,
		);

		expect(screen.getByTestId("previous-button")).toBeDisabled();
	});

	it("should disable next button when canNextPage is false", () => {
		render(
			<Provider>
				<Paginator {...defaultProps} canNextPage={false} />
			</Provider>,
		);

		expect(screen.getByTestId("next-button")).toBeDisabled();
	});

	it("should match snapshot", () => {
		const { asFragment } = render(
			<Provider>
				<Paginator {...defaultProps} />
			</Provider>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
