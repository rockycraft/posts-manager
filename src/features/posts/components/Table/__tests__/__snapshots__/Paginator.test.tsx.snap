// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Paginator component > should match snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="chakra-group css-6l0f26"
  >
    <p
      class="css-0"
      data-first=""
      data-group-item=""
      data-testid="page-detail"
      style="--group-count: 3; --group-index: 0;"
    >
      PÃ¡gina 1 de 5
    </p>
    <button
      class="chakra-button css-4xchnm"
      data-between=""
      data-group-item=""
      data-testid="previous-button"
      style="--group-count: 3; --group-index: 1;"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <button
      class="chakra-button css-4xchnm"
      data-group-item=""
      data-last=""
      data-testid="next-button"
      style="--group-count: 3; --group-index: 2;"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
